modules/standard/Types.chpl

```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
 pragma "no doc"
 inline proc _bxor_id(type t) return 0:t;
 
+
+/* Return the number of fields in a class or record as a param.
+   The count of fields includes types and param fields.
+ */
+proc numFields(type t) param
+  return __primitive("num fields", t);
+
+/* Get the name of the `i`th field in a class or record.
+   `i` should be in 1..numFields(t).
+
+   :arg t: a class or record type
+   :arg i: which field to get the name of
+   :returns: the name of the field, as a param string
+ */
+proc getFieldName(type t, param i:int) param
+  return __primitive("field num to name", t, i);
+
+pragma "no doc"
+proc fieldNumToName(type t, param i:int) param {
+  compilerError("fieldNumToName deprecated. Use getFieldName");
+}
+
+/* Get the `i`th field in a class or record.
+   `i` should be in 1..numFields(t).
+
+   :arg x: a class or record
+   :arg i: which field to get
+   :returns: an rvalue referring to that field.
+ */
+inline
+proc getField(x:?t, param i:int)
+  return __primitive("field value by num", x, i);
+
+// TODO ref return version of getField(x, int)
+
+proc fieldValueByNum(x, param i:int) {
+  compilerError("fieldValueByNum deprecated. Use getField");
+}
+
+/* Get a field in a class or record by name.
+   Will generate a compilation error if a field with that name
+   is not found.
+
+   :arg x: a class or record
+   :arg s: the name of a field
+   :returns: an rvalue referring to that field.
+ */
+proc getField(x:?t, param s:string) {
+  param i = __primitive("field name to num", t, s);
+  if i == 0 then
+    compilerError("field ", s, " not found in ", t:string);
+  return __primitive("field value by num", x, i);
+}
+
+// TODO ref return version of getField(x, string)
+
+/* Get a field index in a class or record.
+   Will not generate a compilation error if the field
+   is not found.
```

Comment:
Would you either change this to describe what it does rather than doesn't do or else remove the sentence and rely on the :returns: clause below (either is fine with me -- I just don't want to get into the practice of enumerating all the things that a routine might do but doesn't b/c there are lots of them!).